plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -1444,7 +1466,9 @@ public void finishMerge(ConnectorSession session, ConnectorMergeTableHandle tabl
                 transactionLogWriter.appendRemoveFileEntry(new RemoveFileEntry(file, writeTimestamp, true));
             }
 
-            List<String> partitionColumns = handle.getMetadataEntry().getOriginalPartitionColumns();
+            List<String> partitionColumns = getColumnMappingMode(handle.getMetadataEntry()) == ColumnMappingMode.NAME
+                    ? getPartitionColumnsForNameMapping(handle.getMetadataEntry().getOriginalPartitionColumns(), ((DeltaLakeMergeTableHandle) tableHandle).getInsertTableHandle().getInputColumns())
```

Comment:
(DeltaLakeMergeTableHandle) tableHandle could be extracted at the beginning of function as it's also used there.